### PR TITLE
ID job performance-tuning practice

### PR DESCRIPTION
**Pull request by grant-guo**

### CODE REVIEW

File: _posts/grant_guo/2017-07-07-spark.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+header:
+    version: small
+    title: Software Engineering Blog
+    image: header-logo-crop.png
+    icon: icon-blog
+---
+
+## Background
+
+In the product environment, DCC-Release ID job takes 20+ hours to finish. The bottleneck is the fact that for every data point ID job needs to talk to DCC-ID service through HTTP request to either obtain or create an ID, which means every time ID job runs there are millions of ID lookup requests, even ID service runs multiple instances and the vanish cache runs on a dedicated server with 128G RAM.
```
Comment (andricDu) [inline _posts/grant_guo/2017-07-07-spark.md]: `through HTTP request` -> `through a HTTP request`

`even ID service runs multiple instances and the vanish cache runs on a dedicated server with 128G RAM.` -> `even if the ID service runs multiple instances, and the vanish cache runs on a dedicated server with 128G RAM.`

File: _posts/grant_guo/2017-07-07-spark.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+## Solution
+
+ID service already has a "****/export" endpoint for each ID type(mutation, sample, donor and specimen) respectively, so the original idea is to cache the exported data at worker nodes locally for lookups and only make the HTTP request when a new ID is required to be generated. 
+
+The first trial is to use Spark Broadcast(https://spark.apache.org/docs/latest/programming-guide.html#broadcast-variables), which is used to give every node a copy of a large input dataset in an efficient manner. 
+
+However I got OutOfMemory exception. I checked the "mutation_ids" table, there are over 80 millions records and each record occupied 1K bytes, which means when loading mutation ID data, approximate 8G memory is needed. After putting more memory to Spark driver, I got another exception which is ConnectionTimedOut. "***/export" endpoint uses Postgresql "COPY" command to export the data. Exporting 8G data and transfering the data through HTTP to Spark driver definitely cost too much, that explains why timeout happened.
```
Comment (andricDu) [inline _posts/grant_guo/2017-07-07-spark.md]: `However I got OutOfMemory exception` -> `However I got OutOfMemory exceptions`

File: _posts/grant_guo/2017-07-07-spark.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+So far Spark broadcast works well for Donor, Sample and Specimen, but failed on Mutation. So the focus is moved to how to load the large dataset of Mutation.
+
+Spark provides a rich set of APIs to load different kinds of data from different data sources. The second trial is to load the data directly from Postgresql through SparkContext.jdbc(...) API. Unfortunately failed again, the size of the dataset is still the root cause. 
+
+At this point, seems everything is stucked. I decided to think about this in a bottom-up way. The whole process, where ID job talks to ID service to obtain or create an ID for each data point, looks like a typical JOIN operation in SQL world, which means I could transform both datasets(from HDFS and Postgresql) to Spark DataFrame, and make a left join operation. If loading Mutation data directly from Postgresql won't work, how about dumping them to HDFS, then loading from HDFS.
```
Comment (andricDu) [inline _posts/grant_guo/2017-07-07-spark.md]: `At this point, seems everything is stucked.` -> `At this point, it seems like everything is stuck.`

**Review by andricDu — CHANGES_REQUESTED** [review]

**Comment by andricDu** [comment]
This is a great post.

**Comment by andricDu** [comment]
You could use this for your teaser image: https://spark.apache.org/images/spark-logo-trademark.png

image

File: _posts/grant_guo/2017-07-07-spark.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+header:
+    version: small
+    title: Software Engineering Blog
+    image: header-logo-crop.png
+    icon: icon-blog
+---
+
+## Background
+
+In our product environment, the DCC-Release ID job takes 20+ hours to finish. The bottleneck is the fact that for every data point ID job needs to talk to DCC-ID service through a HTTP request to either obtain or create an ID, which means every time the ID job runs there are millions of ID lookup requests, even if ID service runs multiple instances, and the vanish cache runs on a dedicated server with 128G RAM.
```
Comment (Fgerthoffert) [inline _posts/grant_guo/2017-07-07-spark.md]: You should break down this paragraph in shorter sentence, suggested version:

The bottleneck is the fact that, for every data point, the ID job needs to talk to DCC-ID service through a HTTP request to either obtain or create an ID. It means every time the ID job runs there are millions of ID lookup requests. This proves to be challenging, even if ID service runs multiple instances and the vanish cache runs on a dedicated server with 128G RAM.

File: _posts/grant_guo/2017-07-07-spark.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+## Solution
+
+The ID service already has a "****/export" endpoint for each ID type(mutation, sample, donor and specimen) respectively, so the original idea is to cache the exported data at worker nodes locally for lookups and only make the HTTP request when a new ID is required to be generated.
```
Comment (Fgerthoffert) [inline _posts/grant_guo/2017-07-07-spark.md]: an ... endpoint (instead of a ... endpoint)

File: _posts/grant_guo/2017-07-07-spark.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+## Background
+
+In our product environment, the DCC-Release ID job takes 20+ hours to finish. The bottleneck is the fact that for every data point ID job needs to talk to DCC-ID service through a HTTP request to either obtain or create an ID, which means every time the ID job runs there are millions of ID lookup requests, even if ID service runs multiple instances, and the vanish cache runs on a dedicated server with 128G RAM.
+
+## Solution
+
+The ID service already has a "****/export" endpoint for each ID type(mutation, sample, donor and specimen) respectively, so the original idea is to cache the exported data at worker nodes locally for lookups and only make the HTTP request when a new ID is required to be generated. 
+
+The first trial is to use Spark Broadcast(https://spark.apache.org/docs/latest/programming-guide.html#broadcast-variables), which is used to give every node a copy of a large input dataset in an efficient manner.
```
Comment (Fgerthoffert) [inline _posts/grant_guo/2017-07-07-spark.md]: replace "which is used to give every node"  by "which gives every node", this also prevent the duplication of "use" in the sentence.

File: _posts/grant_guo/2017-07-07-spark.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+## Solution
+
+The ID service already has a "****/export" endpoint for each ID type(mutation, sample, donor and specimen) respectively, so the original idea is to cache the exported data at worker nodes locally for lookups and only make the HTTP request when a new ID is required to be generated. 
+
+The first trial is to use Spark Broadcast(https://spark.apache.org/docs/latest/programming-guide.html#broadcast-variables), which is used to give every node a copy of a large input dataset in an efficient manner. 
+
+However I got OutOfMemory exceptions. I checked the "mutation_ids" table, there are over 80 millions records and each record occupied 1K bytes, which means when loading mutation ID data, approximate 8G memory is needed. After putting more memory to the Spark driver, I got another exception which is ConnectionTimedOut. "***/export" endpoint uses Postgresql "COPY" command to export the data. Exporting 8G data and transfering the data through HTTP to Spark driver definitely cost too much, that explains why timeout happened.
```
Comment (Fgerthoffert) [inline _posts/grant_guo/2017-07-07-spark.md]: why *the* timeout happened

File: _posts/grant_guo/2017-07-07-spark.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+The ID service already has a "****/export" endpoint for each ID type(mutation, sample, donor and specimen) respectively, so the original idea is to cache the exported data at worker nodes locally for lookups and only make the HTTP request when a new ID is required to be generated. 
+
+The first trial is to use Spark Broadcast(https://spark.apache.org/docs/latest/programming-guide.html#broadcast-variables), which is used to give every node a copy of a large input dataset in an efficient manner. 
+
+However I got OutOfMemory exceptions. I checked the "mutation_ids" table, there are over 80 millions records and each record occupied 1K bytes, which means when loading mutation ID data, approximate 8G memory is needed. After putting more memory to the Spark driver, I got another exception which is ConnectionTimedOut. "***/export" endpoint uses Postgresql "COPY" command to export the data. Exporting 8G data and transfering the data through HTTP to Spark driver definitely cost too much, that explains why timeout happened. 
+
+So far Spark broadcast works well for Donor, Sample and Specimen, but failed on Mutation. So the focus is moved to how to load the large dataset of Mutation.
```
Comment (Fgerthoffert) [inline _posts/grant_guo/2017-07-07-spark.md]: Add comma after "So far"

File: _posts/grant_guo/2017-07-07-spark.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+## Next Step
+
+There are still some potential improvements we could do on ID job and ID service. For example:
+
+ID service's lookup interface is HTTP-based, which is not that efficient as the internal service. We could use RPC solution(say gRPC) with the support of Protocol Buffer or Thrift. 
+
+ID service is using Tomcat as HTTP server, which is the default option from Spring Boot. Although Tomcat has thread pool to handle connections, the basic servlet model is still synchronous. We could use asynchronous web framework like Netty. Futhermore, we could use asynchronous frameworks to access to Postgresql even.
+
+Task scheduling in ID job also could be optimized. So far in the production environment, we have over 70 projects(data sources), 4 types of task(for Donor, Sample, Specimen and Mutation respectively), which means once ID job starts about 70*4 threads contend for the Spark resources because no control of the size of the thread pool. Although our Spark cluster is powerful enough, we could do more fine-grained control.
```
Comment (Fgerthoffert) [inline _posts/grant_guo/2017-07-07-spark.md]: Task scheduling in __the__ ID job

**Review by Fgerthoffert — APPROVED** [review]
Some minor comments, great article.

**Pull request merged**